src/templateScopes.ts

```diff
@@ -0,0 +1,347 @@
+ */
+export class NestedTemplateInnerScope extends TemplateScopeFromObject {
+    public constructor(
+        private nestedTemplateObject: Json.ObjectValue | undefined,
+        // tslint:disable-next-line: variable-name
+        __debugDisplay: string
+    ) {
+        super(
+            nestedTemplateObject,
+            __debugDisplay
+        );
+    }
+
+    public readonly scopeKind: TemplateScopeKind = TemplateScopeKind.NestedDeploymentWithInnerScope;
+
+    protected getResources(): IResource[] | undefined {
+        return getResourcesFromObject(this, this.nestedTemplateObject);
+    }
+}
+
+/**
+ * A nested template with propertes.expressionEvaluationOptions.scope not present or set to "outer".
+ *
+ * Evaluation of expressions, resource() etc. will be in the context of the parent's scope.
+ *
+ * See https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/linked-templates#expression-evaluation-scope-in-nested-templates
+ */
+export class NestedTemplateOuterScope extends TemplateScope {
+    public constructor(
+        public readonly parentScope: TemplateScope,
+        public readonly nestedTemplateObject: Json.ObjectValue | undefined,
+        // tslint:disable-next-line: variable-name
+        __debugDisplay: string
+    ) {
+        super(
+            nestedTemplateObject,
+            __debugDisplay
+        );
+    }
+
+    public readonly scopeKind: TemplateScopeKind = TemplateScopeKind.NestedDeploymentWithOuterScope;
+
+    protected getParameterDefinitions(): IParameterDefinition[] | undefined {
+        return this.parentScope.parameterDefinitions;
+    }
+
+    protected getVariableDefinitions(): IVariableDefinition[] | undefined {
+        return this.parentScope.variableDefinitions;
+    }
+
+    protected getNamespaceDefinitions(): UserFunctionNamespaceDefinition[] | undefined {
+        return this.parentScope.namespaceDefinitions;
+    }
+
+    protected getResources(): IResource[] | undefined {
+        return getResourcesFromObject(this, this.nestedTemplateObject);
+    }
+}
+
+export function getChildTemplateForResourceObject(
+    parentScope: TemplateScope,
+    resourceObject: Json.ObjectValue | undefined // an element of the "resources" section
+): TemplateScope | undefined {
+    // Example nested template
+
+    // "resources": [
+    //     {
+    //         "type": "Microsoft.Resources/deployments",
+    //         "name": "innerScopedNestedTemplate",
+    //         "apiVersion": "2017-05-10",
+    //         "properties": {
+    //             "mode": "Incremental",
+    //             "expressionEvaluationOptions": {
+    //                 "scope": "inner"
+    //             },
+    //             "template": {
+    //                 "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
+    //                 "contentVersion": "1.0.0.0",
+    //                 "parameters": {
+    //                    ..
+
+    // Correct resource type?
```

Comment:
>/ Correct resource type? [](start = 5, length = 24)

Can't you use your isResourceADeployment helper here?

Reply:
Had inlined it, forgot to get rid of isResourceADeployment, done.